model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryMapModelCriteriaBuilder.java

```diff
@@ -0,0 +1,133 @@
+    private final StringBuilder whereClauseBuilder = new StringBuilder(INITIAL_BUILDER_CAPACITY);
+    private final Map<String, Object> parameters = new HashMap<>();
+    private MapModelCriteriaBuilder<K, V, M> mapModelCriteria;
+    public static final Map<SearchableModelField<?>, String> INFINISPAN_NAME_OVERRIDES = new HashMap<>();
+
+    static {
+        INFINISPAN_NAME_OVERRIDES.put(ClientModel.SearchableFields.SCOPE_MAPPING_ROLE, "scopeMappings");
+        INFINISPAN_NAME_OVERRIDES.put(ClientModel.SearchableFields.ATTRIBUTE, "attributes");
+    }
+
+    public IckleQueryMapModelCriteriaBuilder(StringBuilder whereClauseBuilder, MapModelCriteriaBuilder<K, V, M> mapModelCriteria, Map<String, Object> parameters) {
+        this.whereClauseBuilder.append(whereClauseBuilder);
+        this.mapModelCriteria = mapModelCriteria;
+        this.parameters.putAll(parameters);
+    }
+
+    public IckleQueryMapModelCriteriaBuilder(StringKeyConvertor<K> keyConvertor, Class<M> modelClass) {
+        mapModelCriteria = new MapModelCriteriaBuilder<>(keyConvertor, MapFieldPredicates.getPredicates(modelClass));
+    }
+
+    public static String getFieldName(SearchableModelField<?> modelField) {
+        return INFINISPAN_NAME_OVERRIDES.getOrDefault(modelField, modelField.getName());
+    }
+
+    @Override
+    public ModelCriteriaBuilder<M> compare(SearchableModelField<M> modelField, Operator op, Object... value) {
+        StringBuilder newBuilder = new StringBuilder(INITIAL_BUILDER_CAPACITY);
+        newBuilder.append("(");
+
+        if (whereClauseBuilder.length() != 0) {
+            newBuilder.append(whereClauseBuilder).append(" AND (");
+        }
+
+        newBuilder.append(IckleQueryWhereClauses.produceWhereClause(modelField, op, value, parameters));
+
+        if (whereClauseBuilder.length() != 0) {
+            newBuilder.append(")");
+        }
+
+        return new IckleQueryMapModelCriteriaBuilder<>(newBuilder.append(")"), mapModelCriteria.compare(modelField, op, value), parameters);
+    }
+
+    private StringBuilder joinBuilders(ModelCriteriaBuilder<M>[] builders, String delimiter) {
+        return new StringBuilder(INITIAL_BUILDER_CAPACITY).append("(").append(Arrays.stream(builders).map(mcb -> mcb.unwrap(IckleQueryMapModelCriteriaBuilder.class).getWhereClauseBuilder()).collect(Collectors.joining(delimiter))).append(")");
+    }
+
+    private Map<String, Object> joinParameters(ModelCriteriaBuilder<M>[] builders) {
+        return Arrays.stream(builders).map(mcb -> mcb.unwrap(IckleQueryMapModelCriteriaBuilder.class))
+                .map(IckleQueryMapModelCriteriaBuilder::getParameters)
+                .map(params -> (Set<Map.Entry<String, Object>>) params.entrySet())
+                .flatMap(Collection::stream)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    @Override
+    public ModelCriteriaBuilder<M> and(ModelCriteriaBuilder<M>... builders) {
```

Comment:
`or` and `and` methods are probably wrong, I need to take another look and fix them according to the Javadoc